core/src/main/java/org/apache/iceberg/rest/RESTCatalog.java

```diff
@@ -0,0 +1,417 @@
+          clients.first(), tablePath, clients.second(),
+          RESTTableOperations.UpdateType.REPLACE, changes.build(), base);
+
+      return Transactions.replaceTableTransaction(fullName, ops, replacement);
+    }
+
+    @Override
+    public Transaction createOrReplaceTransaction() {
+      // return a create or a replace transaction, depending on whether the table exists
+      // deciding whether to create or replace can't be determined on the service because schema field IDs are assigned
+      // at this point and then used in data and metadata files. because create and replace will assign different
+      // field IDs, they must be determined before any writes occur
+      try {
+        return replaceTransaction();
+      } catch (NoSuchTableException e) {
+        return createTransaction();
+      }
+    }
+
+    private LoadTableResponse stageCreate() {
+      String ns = RESTUtil.urlEncode(ident.namespace());
+      Map<String, String> tableProperties = propertiesBuilder.build();
+
+      CreateTableRequest request = CreateTableRequest.builder()
+          .withName(ident.name())
+          .withSchema(schema)
+          .withPartitionSpec(spec)
+          .withWriteOrder(writeOrder)
+          .withLocation(location)
+          .setProperties(tableProperties)
+          .build();
+
+      // TODO: will this be a specific route or a modified create?
+      return client.post(
+          "v1/namespaces/" + ns + "/stageCreate", request, LoadTableResponse.class, ErrorHandlers.tableErrorHandler());
+    }
+  }
+
+  private static List<MetadataUpdate> createChanges(TableMetadata meta) {
+    ImmutableList.Builder<MetadataUpdate> changes = ImmutableList.builder();
+
+    Schema schema = meta.schema();
+    changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
+    changes.add(new MetadataUpdate.SetCurrentSchema(-1));
+
+    PartitionSpec spec = meta.spec();
+    if (spec != null && spec.isPartitioned()) {
+      changes.add(new MetadataUpdate.AddPartitionSpec(spec));
+      changes.add(new MetadataUpdate.SetDefaultPartitionSpec(-1));
+    }
+
+    SortOrder order = meta.sortOrder();
+    if (order != null && order.isSorted()) {
+      changes.add(new MetadataUpdate.AddSortOrder(order));
+      changes.add(new MetadataUpdate.SetDefaultSortOrder(-1));
+    }
+
+    String location = meta.location();
+    if (location != null) {
+      changes.add(new MetadataUpdate.SetLocation(location));
+    }
+
+    Map<String, String> properties = meta.properties();
+    if (properties != null && !properties.isEmpty()) {
+      changes.add(new MetadataUpdate.SetProperties(properties));
+    }
+
+    return changes.build();
+  }
+
+  private String fullTableName(TableIdentifier ident) {
+    return String.format("%s.%s", catalogName, ident);
+  }
+
+  private static String tablePath(TableIdentifier ident) {
+    return "v1/namespaces/" + RESTUtil.urlEncode(ident.namespace()) + "/tables/" + ident.name();
```

Comment:
What would happen if `ident.name()` contained a slash? I don't believe that there is anything preventing that from happening at the moment.

I checked and a TableIdentifier can be instantiated with a `/` in it. I personally think we should disallow this, as it would likely also cause issues with paths on HDFS.

Reply:
Yeah, this is one of the problems I pointed out when we talked about adding a helper method to generate these paths. This is a test case for that extension.